File: test/scripts/12-many-small-files/incorrect.py
#!/usr/bin/env python3
"""
INCORRECT: Many small files - metadata overhead, inode pressure.
Resume: "Avoid Many Small Files"
"""
import os
import signal
import tempfile
import time
NUM_FILES = 2000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

def main():
    global stop
    tmpdir = tempfile.mkdtemp(prefix='perf-many-')
    print("INCORRECT: 2000 files create/read/delete per iter")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    try:
        while not stop and (time.perf_counter() - start) < DURATION_SEC:
            for i in range(NUM_FILES):
                p = os.path.join(tmpdir, f"{i}.json")
                with open(p, 'w') as f:
                    f.write('{"id":' + str(i) + '}')
            for i in range(NUM_FILES):
                p = os.path.join(tmpdir, f"{i}.json")
                with open(p, 'r') as f:
                    f.read()
            for i in range(NUM_FILES):
                os.unlink(os.path.join(tmpdir, f"{i}.json"))
            laps += 1
    finally:
        for i in range(NUM_FILES):
            try:
                os.unlink(os.path.join(tmpdir, f"{i}.json"))
            except OSError:
                pass
        os.rmdir(tmpdir)
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/12-many-small-files/incorrect.js
const fs = require('fs');
const path = require('path');
const os = require('os');
const NUM_FILES = 2000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), 'perf-many-'));
let laps = 0;
const start = Date.now();
console.log('INCORRECT: 2000 files/iter');
function run() {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    for (let i = 0; i < NUM_FILES; i++) try { fs.unlinkSync(path.join(tmpdir, i + '.json')); } catch(e){}
    fs.rmdirSync(tmpdir);
    console.log('Stopped:', laps, 'laps');
    return;
  }
  for (let i = 0; i < NUM_FILES; i++) {
    fs.writeFileSync(path.join(tmpdir, i + '.json'), '{"id":' + i + '}');
  }
  for (let i = 0; i < NUM_FILES; i++) {
    fs.readFileSync(path.join(tmpdir, i + '.json'));
  }
  for (let i = 0; i < NUM_FILES; i++) {
    fs.unlinkSync(path.join(tmpdir, i + '.json'));
  }
  laps++;
  setImmediate(run);
}
run();
process.on('SIGINT', () => { try { fs.rmdirSync(tmpdir, { recursive: true }); } catch(e){} process.exit(0); });
